Prevent clicks suppressed by osu! resume overlay from getting recorded to replay

bdach opened the pull request:
Intends to close https://github.com/ppy/osu/issues/31758.

I'm not certain about this diff in the slightest but it is an attempt at fixing the issue in question (which, also, by the way, exists in stable). A few things:

- I am not interested in prosecuting specifics of input handling in stable. I looked at stable code for 10 minutes, grabbed my head in desperation, and decided to just invent a behaviour on my own that seems to vaguely make sense and match stable. What stable does I can only describe as "haphazard" but it *seems* that the observable behaviour of it is:

  - Hit circles will ignore new presses from unpausing via [this conditional](https://github.com/peppy/osu-stable-reference/blob/996648fba06baf4e7d2e0b248959399444017895/osu!/GameModes/Play/Player.cs#L3908-L3909)
  - [Sliders](https://github.com/peppy/osu-stable-reference/blob/996648fba06baf4e7d2e0b248959399444017895/osu!/GameplayElements/HitObjects/Osu/SliderOsu.cs#L2017-L2036) and [spinners](https://github.com/peppy/osu-stable-reference/blob/996648fba06baf4e7d2e0b248959399444017895/osu!/GameplayElements/HitObjects/Osu/SpinnerOsu.cs#L256-L266) access the input manager on their own completely bypassing `Unpausing` checks
  
  Therefore the behaviour I want to synthesise here is that hit circles specifically are prohibited from getting activated by an unpause input. Which means that if you want to hit a hovered circle immediately after an unpause then you have to press twice.
- Yes the way this works (~~interdependence on depth~~, weird insertions into ruleset input manager etc.) absolutely blows. No I do not have better ideas.
- I've manually tested that spinners and sliders behave as one would expect. I can add automated testing, after I'm sure this diff doesn't get dumpstered on entry in review.
- In the test scene I've deleted a couple of weirdly specific assertions that I don't know how to fix or whether they even make sense to stay.

Requesting @ppy/team-client review because I want diverse opinions, but also from @frenzibyte specifically because of prior work in this area.

bdach committed:
Add failing test coverage

bdach committed:
Prevent clicks suppressed by osu! resume overlay from getting recorded to replay

bdach committed:
Remove assertions from test that seem to make little sense

peppy commented on osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs:
```diff
@@ -168,18 +176,19 @@ private void updateColour()
             {
                 RelativeSizeAxes = Axes.Both;
-                Depth = float.MinValue;
+                // `ReplayRecorder` specifies `float.MaxValue`. we need to go deeper than it to be able to block inputs from reaching it (depth influences event processing order).
```
```suggestion
                // `ReplayRecorder` specifies `float.MinValue`. we need to go deeper than it to be able to block inputs from reaching it (depth influences event processing order).
```

bdach committed:
Fix incorrect comment

Co-authored-by: Dean Herbert <pe@ppy.sh>

bdach commented on osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs:
```diff
@@ -168,18 +176,19 @@ private void updateColour()
                 RelativeSizeAxes = Axes.Both;
-                Depth = float.MinValue;
+                // `ReplayRecorder` specifies `float.MaxValue`. we need to go deeper than it to be able to block inputs from reaching it (depth influences event processing order).
+                Depth = float.NegativeInfinity;
```
> Would first ask if we need `ReplayRecorded` to have the `Depth` specification in the first place

Well for one this blames back to https://github.com/peppy/osu/commit/467066112f3845393623fa9dc0b2d470a7260935#diff-6b900e931b49c611395cf41e682fb1dbd2f0efa12ce5ff31c5917ae92ddf19edR250, so if anything past you might know that.

The input queue looks like so in osu!:

![Screenshot 2025-05-13 at 11 14 01](https://github.com/user-attachments/assets/06e242d6-374e-4105-95cd-c4a3c790f46b)

The recorder definitely must precede `HitReceptor` in this particular case and there are few good methods of manipulating that that are not `Depth`. And that's only one case, I haven't so much as looked at what the other rulesets do.

peppy replied:
Based on the order there, maybe it's not required anymore? I'd like to remove the specification from `ReplayRecorder` if we can

bdach replied:
It appears that the depth specification *can* be removed from `ReplayRecorder` with no change with respect to gameplay behaviours, so long as you recognise and accept that if we do that, then the ordering here will become *implicit* rather than *explicit*.

As in, at least the current code *attempts* to enforce that the replay recorder should get advance preview of inputs; if the depth set is removed, the fact whether it does get the advance preview of inputs depends on stuff like order of addition to the appropriate drawables and subtree enumeration and such.

I'm not sure I would go as far as to remove it. For one, the resume overlay depth specification *cannot* be removed the same way, unless I try and wedge it via some other method. For another, I am not pushing that change in without specific test coverage that the recorder can e.g. skip inputs because actual gameplay eats them first, as I do not have confidence that is covered by any existing test.

peppy replied:
I'm fine with the one existing in this PR, I just dislike the `Infinite` hack requirement.

We do implicit ordering almost everywhere, so I don't have an issue with this. 🤷

bdach committed:
Merge branch 'master' into resume-press-should-not-be-in-replay

bdach committed:
Remove outdated comment (and use slightly saner value)

bdach committed:
Add test coverage for slider & spinner tracking continuing after resume

bdach committed:
Fix test failures

It's *real* dodgy that this test is sensitive to beatmap conversion
foibles of all things, but I'm not fixing that on this branch because I
don't want review to become outrageously painful due to moving these
tests to requisite ruleset projects.

peppy commented:
@frenzibyte are you intending to look at this one? i'd hope to get this in the next release.

frenzibyte commented:
Was barely able to find time for this as I've been stuck in the song select woods the past few days, I'll try to get to this today.

frenzibyte commented on osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs:
```diff
@@ -178,8 +186,8 @@ public OsuResumeOverlayInputBlocker()
 
             public bool OnPressed(KeyBindingPressEvent<OsuAction> e)
             {
-                bool block = BlockNextPress;
-                BlockNextPress = false;
+                bool block = NextPressToBlock == e.Action;
+                NextPressToBlock = null;
```
Does changing this from boolean to `OsuAction` affect behaviour? or is it just a code quality change.

bdach replied:
I saw it as a reliability change because this component was not apparently ensuring that it's blocking the right input. I don't particularly have an example of the old code being broken but I think the new is safer.

bdach commented on osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs:
```diff
@@ -38,11 +41,9 @@ private void load()
-                var osuPlayfield = drawableOsuRuleset.Playfield;
-                osuPlayfield.AttachResumeOverlayInputBlocker(inputBlocker = new OsuResumeOverlayInputBlocker());
-            }
+            // this relies on execution ordering, particularly on running before `ReplayRecorder` latches onto the `KeyBindingContainer`
+            // (otherwise the press will be recorded to the replay)
```
It is still valid, since the depth spec won't matter if the replay recorder starts recording the replay *before the resume blocker is in the input hierarchy*.

frenzibyte replied:
So it's more of a matter of existing as early as possible rather than strict execution ordering, correct? It doesn't actually matter if the resume blocker is attached just a frame after the replay recorder, or am I missing something?

bdach replied:
I guess? That said this conversation thread feels like prosecuting minor details. I don't think the comment needs changing still.

frenzibyte replied:
Just making sure I'm not missing something, comment is fine.

bdach committed:
Expand & fix tests

bdach committed:
Use until step to see if it helps CI on github actions runners

Was all passing locally...

bdach committed:
Merge branch 'master' into resume-press-should-not-be-in-replay

bdach committed:
Fix tests better

The reason that they were failing on CI but not locally is because of
cd4ab9307c5ed7d160919aa265392ecbad29a138, and because of me forgetting
that CI runs on the merge commit of a PR to master (and correctly so).

The issue is with having presses and releases in a single time instant -
the release will overwrite the press on the replay.

frenzibyte commented on osu.Game.Tests/Visual/Gameplay/TestScenePauseInputHandling.cs:
```diff
@@ -325,32 +372,40 @@ public void TestOsuHitCircleNotReceivingInputOnResume()
-            assertCountOfMatchingReplayFrames<OsuReplayFrame>("left button pressed once in replay", f => f.Actions.Contains(OsuAction.LeftButton), 1);
+            AddUntilStep("left button not pressed again in replay",
+                () => Player.Score.Replay.Frames.OfType<OsuReplayFrame>().Count(f => f.Actions.Contains(OsuAction.LeftButton)),
+                () => Is.Not.GreaterThan(leftFrameCount));
```
The latest change made the test very hard to comprehend. Instead of tracking how many times the button was pressed, this is now tracking how *many* frames exist for the press of a button. This feels quite off.

Can this be swapped with a method that goes through all frames each time and counts how many times the button is pressed after a previous release?

bdach replied:
I don't understand both the concern and what you want me to do here.

The reason that this is "counting frames" is that there is an *indeterminate* number of frames in this test scenario that may contain a pressed button after a press. It may be one, it may be more than one. It's not like the key counter where the counter increments by one every time you press.

bdach committed:
Merge branch 'master' into resume-press-should-not-be-in-replay

bdach committed:
Apply review suggestion

...with adjustments.

smoogipoo commented on osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs:
```diff
@@ -54,9 +55,16 @@ private void load()
+                        if (inputBlocker != null)
+                        {
+                            bool resumeKeyWasNotHeldWhenPausing = !drawableOsuRuleset.AsNonNull().KeyBindingInputManager.PressedActions.Contains(action);
+                            bool mouseOverHittableCircle = drawableOsuRuleset!.Playfield.HitObjectContainer.AliveObjects.Any(h =>
+                                h is DrawableHitCircle && h.IsHovered && !h.IsHit && drawableOsuRuleset.Playfield.HitPolicy.CheckHittable(h, Time.Current, HitResult.None) != ClickAction.Ignore);
```
Do we care about sliders here? Afaict this won't consider them because `AliveObjects` only contains the top-level objects.

bdach replied:
Good spot. I've made it care via https://github.com/ppy/osu/pull/33102/commits/a895958e2430bd46ee0818199f4f4c4f9e36ed6a but I'm not sure what to think of it. It's consistent with hit circles but maybe a touch unfair? But again I'm not sure we should be catering to people who pause at critical points of the map too hard either.

bdach replied:
> maybe it's fine to skip on this discrepancy?

no it's not fine, what do you even mean. the point of this PR is to *stop discrepancies between replay and gameplay*.

peppy replied:
I'm also not sure about this. The primary goal of the resume overlay is to avoid pause buffering, and from the looks of it, this change makes it possible to pause buffer sliders (but not hitcircles). So fixing replay discrepancies at the cost of the resume overlay not working for one hitobject type. On paper this doesn't sound like an amicable solution.

@bdach would you like a fresh set of eyes on this issue at large? I haven't really been involved in the code changes or attempts to fix yet, so maybe I can come up with something that works (low confidence).

bdach committed:
Add failing test coverage for slider heads being treated the same as circles on resume

bdach committed:
Fix tests asserting on incorrect property

bdach committed:
Fix osu! resume overlay blocking logic not considering nested objects

smoogipoo committed:
Merge branch 'master' into resume-press-should-not-be-in-replay

frenzibyte reviewed (commented):
See https://github.com/ppy/osu/pull/33102#discussion_r2170019648

bdach committed:
Revert blocking resume input from reaching slider head circles

It doesn't work because `SliderInputManager` is too complex.

While the resume overlay *can* block the initial press, later things
break down because of

	https://github.com/ppy/osu/blob/e713a68c4936b7f4d9f84d2904dd5e995f7b8820/osu.Game.Rulesets.Osu/Objects/Drawables/SliderInputManager.cs#L255

through which the `SliderInputManager` manages to observe the keypress
it was not supposed to ever see (because the resume overlay was supposed
to be handling it).

I'm not making the `SliderInputManager` logic even more complex to
attempt to bypass this. Even now I'm not sure how I would ever make this
work, that class is way too difficult to make sense of as is anyway.
